File: frameworks/java/spring-boot/UnitTests/src/test/java/unittests/QueryTests.java
package unittests;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import tools.jackson.databind.node.ObjectNode;

class QueryTests extends SpringApp {

    /** query.many's eight values, as a run might draw them. */
    static final Map<String, String> SEARCH = search();

    private static final Set<String> NUMBERS = Set.of("page", "size", "minPrice", "maxPrice");

    // rb:test query.one
    @Test
    @Tag("query.one")
    void oneValueIsBoundAsAnInteger() throws Exception {
        ObjectNode echo = Expected.JSON.createObjectNode().put("page", 417);

        Answer.is(Expected.withEcho("items.small.json", echo), get("/query/one?page=417"));
    }

    // rb:test query.many
    @Test
    @Tag("query.many")
    void eightValuesAreBoundIntoARecord() throws Exception {
        String query = SEARCH.entrySet().stream()
            .map(e -> e.getKey() + "=" + URLEncoder.encode(e.getValue(), StandardCharsets.UTF_8).replace("+", "%20"))
            .collect(Collectors.joining("&"));

        Answer.is(Expected.withEcho("items.small.json", searchEcho()), get("/query/many?" + query));
    }

    /** The echo a handler that bound SEARCH answers with, the numbers as numbers. */
    static ObjectNode searchEcho() {
        ObjectNode echo = Expected.JSON.createObjectNode();
        SEARCH.forEach((name, value) -> {
            if (NUMBERS.contains(name)) {
                echo.put(name, Integer.parseInt(value));
            } else {
                echo.put(name, value);
            }
        });
        return echo;
    }

    private static Map<String, String> search() {
        Map<String, String> values = new LinkedHashMap<>();
        values.put("page", "417");
        values.put("size", "38");
        values.put("status", "paid");
        values.put("category", "garden");
        values.put("sort", "created");
        values.put("q", "alpha bravo");
        values.put("minPrice", "1200");
        values.put("maxPrice", "34000");
        return values;
    }
}

File: frameworks/dotnet/aspnet-mvc/UnitTests/QueryTests.cs
namespace UnitTests;

public sealed class QueryTests(MvcApp app) : IClassFixture<MvcApp>
{
    // rb:test query.one
    [Fact]
    [Trait("corpus", "query.one")]
    public async Task One_value_is_bound_as_an_integer_and_echoed()
    {
        using HttpResponseMessage response = await app.CreateClient().GetAsync("/query/one?page=417");

        await Answer.Is(Expected.WithEcho("items.small.json", new() { ["page"] = 417 }), response);
    }

    // rb:test query.many
    [Fact]
    [Trait("corpus", "query.many")]
    public async Task Eight_values_are_decoded_bound_and_echoed()
    {
        const string query = "page=417&size=38&status=paid&category=garden&sort=created&q=alpha%20bravo&minPrice=1200&maxPrice=34000";

        using HttpResponseMessage response = await app.CreateClient().GetAsync($"/query/many?{query}");

        await Answer.Is(Expected.WithEcho("items.small.json", Search()), response);
    }

    /// <summary>The eight values, as the answer echoes them.</summary>
    internal static JsonObject Search() => new()
    {
        ["page"] = 417,
        ["size"] = 38,
        ["status"] = "paid",
        ["category"] = "garden",
        ["sort"] = "created",
        ["q"] = "alpha bravo",
        ["minPrice"] = 1200,
        ["maxPrice"] = 34000,
    };
}
